### Accuracy:
Some ideas here are taken from the
 [sympy manual](https://docs.sympy.org/latest/modules/evalf.html).

We commented that the operators ``evalf``, ``N``, and ``Float`` have higher precision.
We illustrate this with an example taken from the SymPy manual.

The example illustrates how the 100'th term of the
[Fibonacci series](https://en.wikipedia.org/wiki/Fibonacci_sequence)
, which is
a big number, should be exactly
$$\frac{\varphi^n - (1-\varphi)^n}{\sqrt{5}}.$$
with $\varphi$ equal to the [GoldenRatio](https://en.wikipedia.org/wiki/Golden_ratio).





In [2]:
import sympy as sp
from sympy import GoldenRatio, fibonacci
n=1000
a,b = (GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5), fibonacci(n)
display(float(a))
display(float(b))
float(a)-float(b)

4.3466557686937455e+208

4.3466557686937455e+208

0.0

In [3]:
n=10
sp.N((GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n))

0.e-123

In [4]:
n=1000
sp.N((GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n))

0.e+85

### Actividad 1.
Por que este error tan grande y con `float` no hay error

# `nsimplify`
Muchas de las cosas que aca muestro son del manual
Hemos usado

* `N()`
* `Float()`
* `float()`

Pero hay mas
`nsimplify` convierte aproximaciones en terminos de ecuaciones.
Veamos los ejemplos para mas claridad

In [5]:
from sympy import nsimplify
nsimplify(0.1)

1/10

In [6]:
nsimplify(sp.pi, tolerance=0.01)

22/7

In [7]:
nsimplify(sp.pi, tolerance=0.001)

355/113

In [8]:
# en terminos del GoldenRatio
nsimplify( 4/(1 + sp.sqrt(5)), [GoldenRatio])

-2 + 2*GoldenRatio

In [10]:
# exponencial compleja/real
from sympy import I
nsimplify( I**I, [sp.pi])

exp(-pi/2)

### Actividad 2 : muestre usando variable compleja el anterior resultado

In [12]:
nsimplify( 0.333333, rational='True')

333333/1000000

Observe esta expresion

$$0.333333 = \frac{3}{10} + \frac{3}{10^2} + \cdots \frac{3}{10^6} = 3 \sum_{n=1}^6 \frac{1}{10^n} .$$

Usamos algebra

$$1 + x + x^2 + \cdots x^6 = \frac{1-x^7}{1 -x} .$$




Pasando a restar el 1

$$ x + x^2 + \cdots x^6 = \frac{1-x^7}{1 -x}-1 \frac{1-x^7 - 1 + x }{1 -x} = \frac{x(1-x^6}{1 - x} .$$


Tenemos que $x=1/10$ asi que


$$0.333333 = 3 \frac{10^{-1}(1 - 10^{-6})}{1 - 10^{-1}}= 3 \frac{1 - 1/10^6}{10-1} = 3 \frac{10^6 - 1}{10^6 \times 9}= 3 \frac{999999}{9 \times 10^6} = \frac{333333}{1000000} .$$

Ustedes diran, cual es la gracias


$$0.333333 \times \frac{1000000}{1000000} = \frac{333333}{1000000} .$$

miremos la suma anterior usando `sum`

In [15]:
from sympy import Sum, symbols
k=symbols('k')
s = 3*Sum(1/10**k, (k, 1, 6))
s

3*Sum(10**(-k), (k, 1, 6))

In [16]:
s.evalf()

0.333333000000000

In [17]:
0**0

1

### Actividad 3 : Por que para `python` $0^0=1$?

# Algebra con `SymPy`. Parte I
## monomios, polinomios, etc

In [19]:
from sympy import symbols
x = symbols('x')
x

x

In [20]:
x**2

x**2

In [21]:
# Operador Pow
from sympy import Pow
Pow(x,2)

x**2

In [22]:
3*x

3*x

In [23]:
# otra forma de importar symbolos  usando `abc`
from sympy.abc import x,y,z,w
f = x**2 + y - z + w
f

w + x**2 + y - z

## Polinomios
### suma y resta

In [24]:
y = x**2 - x - 6 # suma de monomios
y

x**2 - x - 6

In [25]:
w = x**3 - x + 8
y+w

x**3 + x**2 - 2*x + 2

In [26]:
# usamos el simbolo `ADD`
from sympy import Add
Add(y,w)

x**3 + x**2 - 2*x + 2

In [27]:
y-w  # resta

-x**3 + x**2 - 14

## Multiplicacion y Factorizacion

In [28]:
z = (x+5)*(x-2)
z

(x - 2)*(x + 5)

In [29]:
# expandimos
from sympy import expand
zexpand = expand(z)
zexpand

x**2 + 3*x - 10

In [30]:
display(y*w)
h=(y*w).expand()
h

(x**2 - x - 6)*(x**3 - x + 8)

x**5 - x**4 - 7*x**3 + 9*x**2 - 2*x - 48

In [32]:
# usando MUL
from sympy import Mul
display(Mul(y,w))
Mul(y,w).expand()

(x**2 - x - 6)*(x**3 - x + 8)

x**5 - x**4 - 7*x**3 + 9*x**2 - 2*x - 48

In [33]:
# combinamos operadores
Add( Pow(x,2), Mul(x,y))

x**2 + x*(x**2 - x - 6)

## La binomial

In [35]:
a = symbols('a')
b = symbols('b')

expand( (a+b)**5)

a**5 + 5*a**4*b + 10*a**3*b**2 + 10*a**2*b**3 + 5*a*b**4 + b**5

### Factorizacion<

In [37]:
from sympy import factor
zfactor = factor(zexpand)
zfactor

(x - 2)*(x + 5)

In [38]:
factor(y)

(x - 3)*(x + 2)

In [39]:
h.factor()

(x - 3)*(x + 2)*(x**3 - x + 8)

In [40]:
h2 = x**3 - x + 8
factor(h2, extension=[I])

x**3 - x + 8

#### factorizacion sobre los complejos.

Como factorizamos

$$x^2 + 1$$?
Este polinomio (binomio) es **irreducible** sobre los numeros reales. Pero se puede **factorizar** sobre los complejos.

Sea $\mathrm{i}=\sqrt{-1}$.

Se puede verificar, facilmente que


$$ x^2 + 1 = (x - \mathrm{i})(x + \mathrm{i}).$$

In [42]:
factor(x**2 + 1, extension=[I])

(x - I)*(x + I)

### Actividad 4
Por que `SymPy` no factorizo $x^3 - x + 8$ sobre los complejos?

In [44]:
from sympy import roots
r = roots(h2)
display(h2)
display(r)

x**3 - x + 8

{-(3*sqrt(1293) + 108)**(1/3)/3 - 1/(3*sqrt(1293) + 108)**(1/3): 1,
 -(-1/2 + sqrt(3)*I/2)*(3*sqrt(1293) + 108)**(1/3)/3 - 1/((-1/2 + sqrt(3)*I/2)*(3*sqrt(1293) + 108)**(1/3)): 1,
 -1/((-1/2 - sqrt(3)*I/2)*(3*sqrt(1293) + 108)**(1/3)) - (-1/2 - sqrt(3)*I/2)*(3*sqrt(1293) + 108)**(1/3)/3: 1}

### Collect
Considere, por ejemplo,

$$ xy + xz + x^2 y z + yz .$$
Por alguna razon queremos expresar este como un polinomio en variable $x$. Es decir


$$p(x) = yz + x(y+z) + x^2 yz .$$
Veamos como `SymPy` nos da lo que queremos

In [45]:
from sympy import symbols, collect
x,y,z = symbols('x y z')
expr = x*y + x*z + x**2 * y * z + y*z
display(expr)
coll = collect(expr, x)
coll

x**2*y*z + x*y + x*z + y*z

x**2*y*z + x*(y + z) + y*z

In [49]:
coll.coeff(x,2)

y*z

### Simplificacion con radicales (esto tal vez pertenece al topico de Aritmetica


$$\frac{1}{1 - \sqrt{2}} = \frac{2 + \sqrt{2}}{(2- \sqrt{2})(2 + \sqrt{2})} = \frac{2 + \sqrt{2}}{2} .$$

In [51]:
from sympy import radsimp, sqrt
radsimp(1/(2 - sqrt(2)))

(sqrt(2) + 2)/2

### Manejo de "strings" en `SymPy`.  
Uso de `srepr()`
El uso de `eval`

In [52]:
import sympy as sp
from sympy import diff # esto es derivada, topico de calculo, por ahora lo hacemos con fuerza bruta
a,b,c,s,x = symbols('a b c s x')
s=" a + sp.exp(-b*(x-c)**2)" # es una cadena "string"
diff( eval(s), x)

-b*(-2*c + 2*x)*exp(-b*(-c + x)**2)

In [58]:
# en el sentido contrario
from sympy import srepr # string representation
srepr(diff(eval(s),x))

"Mul(Integer(-1), Symbol('b'), Add(Mul(Integer(-1), Integer(2), Symbol('c')), Mul(Integer(2), Symbol('x'))), exp(Mul(Integer(-1), Symbol('b'), Pow(Add(Mul(Integer(-1), Symbol('c')), Symbol('x')), Integer(2)))))"